### PR TITLE
En 3698 termui update new metrics

### PR DESCRIPTION
**Pull request by miiu96**
add new termui metrics :
- synchronization estimation
- block size 
- number of processed txs
- current block hash
- current round timestamp

### CODE REVIEW

File: consensus/chronology/chronology_test.go [hunk 1]
```diff
@@ -325,7 +325,7 @@ func TestChronology_SetAppStatusHandlerWithOkValueShouldPass(t *testing.T) {
 func TestChronology_CheckIfStatusHandlerWorks(t *testing.T) {
 	t.Parallel()
 
-	chanDone := make(chan bool, 1)
+	chanDone := make(chan bool, 2)
```
Comment (sasurobert) [inline consensus/chronology/chronology_test.go]: what does 2 means ?

File: process/block/metablock.go [hunk 2]
```diff
@@ -155,6 +155,8 @@ func (mp *metaProcessor) ProcessBlock(
 		return process.ErrWrongTypeAssertion
 	}
 
+	go mp.getMetricsFromMetaHeader(*header)
```
Comment (sasurobert) [inline process/block/metablock.go]: send via pointer. getMetricsFromMetaheader(header)

File: process/block/metablock.go [hunk 3]
```diff
@@ -211,6 +213,32 @@ func (mp *metaProcessor) ProcessBlock(
 	return nil
 }
 
+func (mp *metaProcessor) getMetricsFromMetaHeader(header block.MetaBlock) {
+	numMiniBlocksMetaBlock := uint64(0)
+	headerSize := uint64(0)
+
+	for _, shardInfo := range header.ShardInfo {
+		numMiniBlocksMetaBlock += uint64(len(shardInfo.ShardMiniBlockHeaders))
+
```
Comment (sasurobert) [inline process/block/metablock.go]: delete empty line

File: process/block/shardblock.go [hunk 4]
```diff
@@ -257,6 +251,35 @@ func (sp *shardProcessor) ProcessBlock(
...
+		if err == nil {
+			miniblocksSize += uint64(len(marshalizedBlock))
+		}
+	}
+	sp.appStatusHandler.SetUInt64Value(core.MetricNumTxInBlock, uint64(totalTxCount))
+	sp.appStatusHandler.SetUInt64Value(core.MetricNumMiniBlocks, uint64(mbLen))
+	sp.appStatusHandler.SetUInt64Value(core.MetricMiniBlocksSize, miniblocksSize)
+}
+
+func (sp *shardProcessor) getMetricsFromHeader(header block.Header, numTxWithDst uint64, totalTx int) {
```
Comment (sasurobert) [inline process/block/shardblock.go]: use pointer

File: process/block/shardblock.go [hunk 5]
```diff
@@ -257,6 +251,35 @@ func (sp *shardProcessor) ProcessBlock(
 	return nil
 }
 
+func (sp *shardProcessor) getMetricsFromBlockBody(body block.Body) {
```
Comment (sasurobert) [inline process/block/shardblock.go]: do this in metrics.go - add these code there.

File: statusHandler/presenter/common.go [hunk 6]
```diff
@@ -30,3 +35,45 @@ func (psh *PresenterStatusHandler) getFromCacheAsString(metric string) string {
...
+	if (count == 1) || (count == 0) {
+		result = strconv.Itoa(count) + " " + singular + " "
+	} else {
+		result = strconv.Itoa(count) + " " + singular + "s "
+	}
+	return
+}
+
+func secondsToHuman(input int) (result string) {
+	years := math.Floor(float64(input) / 60 / 60 / 24 / 7 / 30 / 12)
```
Comment (sasurobert) [inline statusHandler/presenter/common.go]: refactor this code with less magic numbers.

File: statusHandler/presenter/common.go [hunk 7]
```diff
@@ -30,3 +35,45 @@ func (psh *PresenterStatusHandler) getFromCacheAsString(metric string) string {
 
 	return valStr
 }
+
+func plural(count int, singular string) (result string) {
+	if (count == 1) || (count == 0) {
```
Comment (sasurobert) [inline statusHandler/presenter/common.go]: if count < 2, no need for else as you can return inside the if

File: statusHandler/presenter/chainInfoGetters.go [hunk 8]
```diff
@@ -51,3 +55,89 @@ func (psh *PresenterStatusHandler) GetNumConnectedPeers() uint64 {
 func (psh *PresenterStatusHandler) GetCurrentRound() uint64 {
 	return psh.getFromCacheAsUint64(core.MetricCurrentRound)
 }
+
+// CalculateSynchronizationEstimation will calculate and return an estimation of
+// the time required for synchronization in a human friendly format
+func (psh *PresenterStatusHandler) CalculateSynchronizationEstimation() string {
```
Comment (sasurobert) [inline statusHandler/presenter/chainInfoGetters.go]: rename function. CalculateTimeToSynchronize

File: statusHandler/presenter/chainInfoGetters.go [hunk 9]
```diff
@@ -51,3 +55,89 @@ func (psh *PresenterStatusHandler) GetNumConnectedPeers() uint64 {
...
+	currentBlock := psh.GetNonce()
+
+	psh.mutEstimationTime.Lock()
+	diffTime := time.Now().Sub(psh.startTime).Seconds()
+	blocksSynchronized := currentBlock - psh.startBlock
+	psh.mutEstimationTime.Unlock()
+
+	if blocksSynchronized == 0 {
```
Comment (sasurobert) [inline statusHandler/presenter/chainInfoGetters.go]: I think an estimation could be calculated even if no blocks were synchronized

File: statusHandler/presenter/chainInfoGetters.go [hunk 10]
```diff
@@ -51,3 +55,89 @@ func (psh *PresenterStatusHandler) GetNumConnectedPeers() uint64 {
...
+}
+
+// CalculateSynchronizationSpeed will calculate and return speed of synchronization
+// how many block per second are synchronized
+func (psh *PresenterStatusHandler) CalculateSynchronizationSpeed() uint64 {
+	currentBlock := psh.GetNonce()
+	psh.mutEstimationTime.Lock()
+	blocksSynchronized := currentBlock - psh.startBlock
+	psh.mutEstimationTime.Unlock()
+	if blocksSynchronized == 0 {
```
Comment (sasurobert) [inline statusHandler/presenter/chainInfoGetters.go]: I think an estimation could be calculated even if no blocks were synchronized

File: statusHandler/presenter/chainInfoGetters.go [hunk 11]
```diff
@@ -51,3 +55,89 @@ func (psh *PresenterStatusHandler) GetNumConnectedPeers() uint64 {
...
+		blocksPerSecond = 0
+	}
+
+	psh.oldNonce = currentBlock
+
+	return uint64(blocksPerSecond)
+}
+
+// PrepareForCalculationSynchronizationTime prepare information that are need to calculate synchronization time
+func (psh *PresenterStatusHandler) PrepareForCalculationSynchronizationTime() {
```
Comment (sasurobert) [inline statusHandler/presenter/chainInfoGetters.go]: redo this function. It is an infinite loop. not ok. You can calculate the synchronization time need for one block in other ways. You get all the needed data, when the block is changed in one of the setters you can calculate the time elapsed since the last block. No need for this function.

File: core/converters.go [hunk 12]
```diff
@@ -15,7 +15,7 @@ func ConvertBytes(bytes uint64) string {
 		return fmt.Sprintf("%d B", bytes)
 	}
 	if bytes < 1024*1024 {
-		return fmt.Sprintf("%.2f kiB", float64(bytes)/1024.0)
+		return fmt.Sprintf("%.2f KB", float64(bytes)/1024.0)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply core/converters.go]: further more down we have MB, GB, why use kB?

File: cmd/node/main.go [hunk 13]
```diff
@@ -715,15 +715,22 @@ func initMetrics(
 	appStatusHandler.SetStringValue(core.MetricNodeType, string(nodeType))
 	appStatusHandler.SetUInt64Value(core.MetricRoundTime, roundDuration/milisecondsInSecond)
 	appStatusHandler.SetStringValue(core.MetricAppVersion, version)
-	appStatusHandler.SetUInt64Value(core.MetricCountConsensus, 0)
-	appStatusHandler.SetUInt64Value(core.MetricCountLeader, 0)
-	appStatusHandler.SetUInt64Value(core.MetricCountAcceptedBlocks, 0)
-	appStatusHandler.SetUInt64Value(core.MetricNumTxInBlock, 0)
+	appStatusHandler.SetUInt64Value(core.MetricCountConsensus, uint64(0))
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/main.go]: Done.

File: cmd/node/main.go [hunk 14]
```diff
@@ -715,15 +715,22 @@ func initMetrics(
...
+	appStatusHandler.SetUInt64Value(core.MetricCountLeader, uint64(0))
+	appStatusHandler.SetUInt64Value(core.MetricCountAcceptedBlocks, uint64(0))
+	appStatusHandler.SetUInt64Value(core.MetricNumTxInBlock, uint64(0))
 	appStatusHandler.SetUInt64Value(core.MetricNumMiniBlocks, 0)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/main.go]: Done.

File: consensus/chronology/chronology_test.go (hunk 1 above)
Comment (miiu96) [reply consensus/chronology/chronology_test.go]: We need a buffered channel because channel is set 2 times in test.

File: process/block/metablock.go (hunk 3 above)
Comment (miiu96) [reply process/block/metablock.go]: Done.

File: process/block/shardblock.go (hunk 5 above)
Comment (miiu96) [reply process/block/shardblock.go]: Done. Methods that set metrics was moved in a new file.

File: statusHandler/presenter/chainInfoGetters.go (hunk 8 above)
Comment (miiu96) [reply statusHandler/presenter/chainInfoGetters.go]: Done.

File: statusHandler/presenter/chainInfoGetters.go (hunk 9 above)
Comment (miiu96) [reply statusHandler/presenter/chainInfoGetters.go]: The estimate is calculated in another way.

File: statusHandler/presenter/chainInfoGetters.go (hunk 10 above)
Comment (miiu96) [reply statusHandler/presenter/chainInfoGetters.go]: Another implementation for estimation was written.

File: statusHandler/presenter/chainInfoGetters.go (hunk 11 above)
Comment (miiu96) [reply statusHandler/presenter/chainInfoGetters.go]: The function has been removed. No more need of it.

File: cmd/node/main.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/main.go]: These cast are needed because we want to set and uint64 value.

File: process/block/metablock.go [hunk 15]
```diff
@@ -211,6 +213,32 @@ func (mp *metaProcessor) ProcessBlock(
...
+		headerSize = uint64(len(marshalizedHeader))
+	}
+
+	mp.appStatusHandler.SetUInt64Value(core.MetricHeaderSize, headerSize)
+	mp.appStatusHandler.SetUInt64Value(core.MetricNumTxInBlock, uint64(header.TxCount))
+	mp.appStatusHandler.SetUInt64Value(core.MetricNumMiniBlocks, numMiniBlocksMetaBlock)
+
+	numShardHeadersFromPool := 0
+	shardMBHeaderCounterMutex.Lock()
+	mp.appStatusHandler.SetUInt64Value(core.MetricNumShardHeadersProcessed, uint64(shardMBHeadersTotalProcessed))
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/block/metablock.go]: This will be done in another pull request. Already task with refactoring metablock.go is in progress.

File: process/block/metablock.go [hunk 16]
```diff
@@ -1173,14 +1201,15 @@ func (mp *metaProcessor) displayLogInfo(
 	}
 
 	shardMBHeaderCounterMutex.RLock()
+	headerHashS := core.ToB64(headerHash)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/block/metablock.go]: Variable was renamed with a suggestive name.

File: process/block/metablock.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply process/block/metablock.go]: Done.

File: statusHandler/presenter/common.go (hunk 7 above)
Comment (miiu96) [reply statusHandler/presenter/common.go]: Done.

File: process/block/shardblock.go (hunk 4 above)
Comment (miiu96) [reply process/block/shardblock.go]: Done.

File: process/block/metablock.go (hunk 2 above)
Comment (miiu96) [reply process/block/metablock.go]: Done.

File: statusHandler/presenter/common.go [hunk 17]
```diff
@@ -30,3 +35,45 @@ func (psh *PresenterStatusHandler) getFromCacheAsString(metric string) string {
...
+		result = strconv.Itoa(count) + " " + singular + " "
+	} else {
+		result = strconv.Itoa(count) + " " + singular + "s "
+	}
+	return
+}
+
+func secondsToHuman(input int) (result string) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/presenter/common.go]: Function was refactored.

File: statusHandler/presenter/common.go (hunk 6 above)
Comment (miiu96) [reply statusHandler/presenter/common.go]: Done.

File: core/converters.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply core/converters.go]: ok KB

File: core/converters.go [hunk 18]
```diff
@@ -60,3 +62,37 @@ func CalculateHash(
...
+		result = strconv.Itoa(count) + " " + singular + "s "
+	}
+	return
+}
+
+// SecondsToHumanFormat transform seconds input in a human friendly format
+func SecondsToHumanFormat(input int) string {
+	numSecondsInAMinute := 60
+	numMinutesInAHour := 60
+	numSecondsInAHours := numSecondsInAMinute * numMinutesInAHour
```
Comment (iulianpascalau) [inline core/converters.go]: typo, Hours->Hour

File: core/converters.go [hunk 19]
```diff
@@ -60,3 +62,37 @@ func CalculateHash(
...
+	if count < 2 {
+		result = strconv.Itoa(count) + " " + singular + " "
+	} else {
+		result = strconv.Itoa(count) + " " + singular + "s "
+	}
+	return
+}
+
+// SecondsToHumanFormat transform seconds input in a human friendly format
+func SecondsToHumanFormat(input int) string {
```
Comment (iulianpascalau) [inline core/converters.go]: This is not a Format it's a conversion. Should rename to something like SecondsToHuman or SecondsToHourMinSec

Comment (iulianpascalau) [inline core/converters.go]: This is not tested. Add some tests with some edge cases.

Comment (miiu96) [reply core/converters.go]: Done. Name was changed.

Comment (miiu96) [reply core/converters.go]: Done.

File: core/converters.go (hunk 18 above)
Comment (miiu96) [reply core/converters.go]: Done.

File: core/converters_test.go [hunk 20]
```diff
@@ -82,3 +82,38 @@ func TestCalculateHash_Good(t *testing.T) {
 	assert.Nil(t, err)
 	assert.True(t, bytes.Equal(results, hash))
 }
+
+func TestSecondsToHourMinSec_ShouldWork(t *testing.T) {
```
Comment (iulianpascalau) [inline core/converters_test.go]: :+1:
